### ISSUE TITLE
Make Mint Loader version independent

### ISSUE THREAD

**Issue by ps68060** [freemint/freemint#138]
It would be useful if the Mint loader in the AUTO folder was independent of the (minor) version.

Perhaps it could check the Mint folder for available versions and let the user decide which to start.

The last version used could be stored (mint.ini  ?)

**Comment by mikrosk** [comment]
The idea of this (my) tool has been explicitly that - to load the right CPU version for given version. On the other hand, I have never thought of this use case before. It's pretty simple to bundle the [per-CPU snapshots](https://bintray.com/freemint/freemint/snapshots-cpu/_latestVersion) with mint loader in `AUTO` and still let it load `mint0?0.prg` according to the current CPU but always from `1-19-cur`.

The only question then is that for what CPU should `mintloader.prg` be optimised for in every archive (for instance, the 060 archive has everything built for 060 however you may want to side-boot it in 030 mode and it could throw an unsupported instruction exception).

**Comment by mikrosk** [comment]
Completely different question is to extend mintloader into something more, i.e. proving a range of kernels to boot in some interactive way for example. My gut feeling was always to let this to a boot manager but I realise mintloader already does some of its tasks.

**Comment by ps68060** [comment]
My thought was something simple to start with that avoids having to update the mint loader every time.

Move mint.ini to a more general location such as c:\AUTO or c:\MINT
Add a line to minit.ini to specify the version or folder to load mint from.

Later on it could be improved by adding to the mint boot menu to allow version selection.

**Comment by mikrosk** [comment]
My use case was to have 2-3 MINT0?0.PRGs in `AUTO` and with every update I had to update / delete / rename all of them, what was really annoying.

The CPU builds do not contain mint loader in AUTO at all, they contain MINT0?0.PRG only.

So if you're after archives which you can just update, the CPU builds are for you. As I have mentioned, I'm open to add mint loader in those archives as well, if you find it helpful (i.e. you wouldn't need to delete `MINT040.PRG` in AUTO in case you are using only `MINTMIL.PRG`).

I'd prefer to keep preconfigured / generic builds as they are because they are super easy to test - if anything goes wrong, you basically just swap/delete the mint loader binary. With your proposal I'd need to edit `MINT.INI`, rename folders in `C:\MINT` (what is not always possible when running the kernel from there) etc. And testing new kernels I always consider as priority, because now I can ask anyone just to unpack the archive on C:, test something for me and delete it, without messing with anything else.

**Comment by th-otto** [comment]
> The only question then is that for what CPU should mintloader.prg be optimised for

Wasn't that the main goal for mintloader, to have only one version of it, and load the mint kernel for the current CPU type?

**Comment by ps68060** [comment]
Mikro,

I think you've misunderstood.

No folder renaming needed.
No changing the mint loader in AUTO.

The problem with the current system is that the loader in AUTO has to be updated every time and we have to be careful to preserve the boot order.

My idea is to remove the hard-coded version in the AUTO loader use config information in mint.ini to load the version. This means that the loader can be kept the same most of the time.

**Comment by mikrosk** [comment]
> Wasn't that the main goal for mintloader, to have only one version of it, and load the mint kernel for the current CPU type?

Yes but currently it's always in the compatible "group" - so if mintloader is compiled for and the archive is targeted for `-m68020-60`, it's totally fine that it chooses a 020, 030, 040 or 060 kernel. However if one downloads pure 060 archive (which contains only 060 kernels), it's a bit different because it can be still booted in CT60's 030 mode, for instance. But that is easily solvable by forcing it to always compile for `-m68000`, sure.

@ps68060 no no, I understand. However I think you have missed my point - if you are talking about *updating*, you shouldn't be downloading the preconfigured archives, you should be downloading the CPU archives, as mentioned in my first post.

**Comment by ps68060** [comment]
The archive naming doesn't seem to describe what they are.

I download the CPU named archives and they are a different build version so the AUTO prg has to be replaced.

Why is the loader linked to the version ?  If it was changed to load any available build then the user can choose at boot time to switch from 1-19-x to 1-19-y

**Comment by mikrosk** [comment]
> I download the CPU named archives and they are a different build version so the AUTO prg has to be replaced.

Yes. And if *this* is the problem you are trying to solve, this is where my proposal comes from, i.e. I could include one single mintloader (which wouldn't have to be replaced, ever) and you would be just replacing `C:\MINT\1-19-CUR` (so no `AUTO` update would be needed anymore). But only in CPU builds.

> Why is the loader linked to the version ? If it was changed to load any available build then the user can choose at boot time to switch from 1-19-x to 1-19-y

Because the minor version = commit ID. About choosing, see above, that requires a lot of code, a lot of decision and I'm absolutely sure someone would be against some of them therefore I really don't feel up this, especially when we consider making a release in "near" future.

**Comment by ps68060** [comment]
That's why I suggest a simple first step of just putting the version in mint.ini rather than hard-coding.

Then users testing the CI builds only have to edit the file rather than changing AUTO contents. Yes the random CI build names are bit inconvenient.

I am trying to avoid the fiddling that I do at the moment...

Currently in my AUTO folder I have 2 mint loaders:

e.g.
1. "current" build which is named mint-new.prg (this is so I don't have to remember which is the latest)
2. a previous build with it's real name eg. mint-120.prx

To update I rename the old mint-120.prx to match the newly downloaded version. eg. mint-123.prx.
Then over-write that file with the Actual mint-123.prx 
This safely preserves the entire boot order.

Then when I boot I use X-Boot to choose either mint-new or mint-123

Once I am happy that the latest build is good I rename mint-123 to make it mint-new

**Comment by mikrosk** [comment]
Wait, now you have really lost me. So you are both happy with being able to test new builds without destroying your old setup and at the same time you would like to have mint loader version independent. I understand that your idea includes extending it to handling of `mint.ini` but `mint.ini` is version dependent. So those are the decisions I'm talking about, I am personally happy with `mint.ini` being local to each version.

To me it looks like you don't the basic trick of preserving `AUTO` order files: just replace the old one. So you have `mint-ok.prx` (old) and `mint-124.prg` (new), you are happy with the 124, so press CONTROL, drag & drop `mint-124.prg` to the same folder, you get asked for duplicate filename (124),  rename it to ok, again asked for duplicate name, press OK to overwrite, done.

**Comment by mikrosk** [comment]
Actually, maybe the second part of my comment is exactly what you meant with that update/rename procedure but then again, I'm really confused what we are trying to solve here.

**Comment by ps68060** [comment]
I described the "trick" of replacing the old one with the same filename, above.

I maintain 2 mint loader versions in AUTO just in case the new one doesn't boot for some reason. I have had this happen more than once.

My process above maintains the order when 2 versions of mint loader.

However, if mint loader was able to load any version then it need  not be replaced for long periods. Using it rather like grub, where it can be used to select from a list of builds.

**Comment by mikrosk** [comment]
Ah ok, I finally get it. So you are literally after a feature for loading the version from a config file and being able to add one, two, three, ... kernels while still keeping the correct boot order.

As a workaround I can still imagine having `1-19-cur` mint loader and renaming folders in `C:\MINT` (i.e. `1-19-cur` would be your current setup and `1-19-cur.old` would be your inactive one). Not as comfortable as your proposal but certainly easier to do.

**Comment by ps68060** [comment]
My thought was to split the dev into 2 phases.

Phase 1 use mint.ini and then in phase 2 check the folders at boot time and offer the user a list to pick from.

**Comment by paulwratt** [comment]
.. and the `c:\mint\mint.ini` content could be used to manage a timeout. Or rather use `mintload.ini` to avoid confusion, because the MiNT INI can (and has) change between versions, including sub-versions.

If i remember correctly the kernel loaders check 3 different folders, first is the release build version including sub-version (`C:\MINT\1-16-3`), next is current for that major build version (`C:\MINT\1-16-cur`), and there is a 3rd falback folder I cant remember (unrelated to versions?), that could be used to advantage by _mintloader_.

Plus any folder in `C:\MINT\?` that starts with a number is obviously a viable kernel version, and if they contain an `AUTO` folder, you have access to any kernels present, based on CPU and/or machine.

----

**FYI**

I had similar problem to both @ps68060 and @mikrosk for ARAnyM, but worse (not just which sub-version to load as this thread discusses):
 - aranym.inf
 - ROM
 - MiNT
 - VDI
 - resolution
 - AES
 - theme
all of which may be changed independantly. I had MiNT versions `1-16-3`, `1-16-cur`, `1-17-cur-good`, `1-17-cur-cur` and `1-17-cur-test`. At the time I did not know you could also boot from the auto-folder with ARAnyM, just like real hardware, until I found MiniPAK.

I used `C:\MINT` to contain the static file names that were referenced from all configs, and copied the correct file or folder to the static name: `ARANYM.INF`, `ARANYM.VER`, `ARANYM.ROM`, `ROM.VER`, `MINTARA.PRG`, `1-16-CUR`, `1-16.VER`, `1-17-CUR`, `1-17.VER`, `MINT.VER`, `FVDI.VER`, `FVDI.REZ`, `VDI.VER`, `XAAES.CNF`, `XAAES.VER`, `MYAES.INF`, `MYAES.VER`, `AES.VER`, `DESKTOP.INF`, `DESKTOP.VER`, `THEME.VER`, `LANG.VER`, `PREVIOUS.VER`, `BOOT.VER`, etc.

I may have put them in a sub folder `C:\MINT\BOOT`, cant remember exactly, but you could easily tell what the current and previous configuration was, either via the individual text files, or the configur8r script without options, where **_BOOT_** and **_PREVIOUS_** contained condensed information from the other files (which was also useful to iscolate both good & bad configurations).

I combined that static naming system with a (configur8r) shell script that could tell which side of ARAnyM it was being run from (inside or outside), and what was available as options to change. I never got to write the GEM Configur8r, but Helmut did add the ability to include external CNF files so I could easily change themes, VDI and AES etc.

The same filesystem structure was intended to be usable from various emulators, specifically Hatari TT 68000 extended resolution, SS-STEem, MiniPAK, and AFROS-LiveCD, with SpareMiNT, EasyMiNT, GentooMiNT and "TOSMiNT". 

BTW - "TOSMiNT" is just a specific AFROS structure, TOS/GEM only, no ext2 filesystem, but with seperate full Bash (260K/800K?) and (default) mini Bash (42K, no internals) and _full_ _condensed_ `bin` in `u:/c/mint/bin` and equivalent (useful) `c:\mint\ttp` (or `c:\mint\tos`), with `tools` in `c:\mint\tools`. The Bash and `bin` & `ttp` was available via AFROS-update.

I know most of this **FYI** is not directly related, but it might give ideas and insparation as to automation, multiple platforms and environments, especially where multiple MiNT versions are used, needed or tested, and/or where the filesystem is used on multple targets (eg CF-Card).
